File: server/app/analysis.py
"""Local Stockfish analysis, used as a fallback for games Lichess hasn't
already analyzed (most bullet/blitz games, since server-side analysis on
lichess.org is opt-in per game)."""
import shutil

import chess
import chess.engine

from app.config import STOCKFISH_DEPTH, STOCKFISH_PATH, STOCKFISH_TIME_LIMIT

# Debian/Ubuntu's stockfish package installs to /usr/games, which isn't
# always on PATH (e.g. non-login shells, some containers).
_FALLBACK_PATHS = ["/usr/games/stockfish", "/usr/local/bin/stockfish", "/opt/homebrew/bin/stockfish"]


class EngineUnavailable(Exception):
    pass


def resolve_stockfish_path() -> str | None:
    found = shutil.which(STOCKFISH_PATH)
    if found:
        return found
    for candidate in _FALLBACK_PATHS:
        if shutil.which(candidate):
            return candidate
    return None


def stockfish_available() -> bool:
    return resolve_stockfish_path() is not None


def classify_loss(loss_cp: int) -> str:
    if loss_cp >= 300:
        return "blunder"
    if loss_cp >= 120:
        return "mistake"
    if loss_cp >= 40:
        return "inaccuracy"
    if loss_cp <= 5:
        return "best"
    return "good"


def analyze_moves_san(moves_san: list[str], clocks_centis: list[int] | None = None) -> list[dict]:
    """Runs Stockfish over every position of a game and returns per-move
    eval + classification. `moves_san` is the game's move list in SAN."""
    resolved_path = resolve_stockfish_path()
    if not resolved_path:
        raise EngineUnavailable(
            f"Stockfish binary not found ('{STOCKFISH_PATH}'). "
            "Install it (e.g. `apt-get install stockfish`) or set STOCKFISH_PATH."
        )

    limit = (
        chess.engine.Limit(time=STOCKFISH_TIME_LIMIT)
        if STOCKFISH_TIME_LIMIT > 0
        else chess.engine.Limit(depth=STOCKFISH_DEPTH)
    )

    board = chess.Board()
    relative_cp: list[int] = []
    best_moves_san: list[str | None] = []

    with chess.engine.SimpleEngine.popen_uci(resolved_path) as engine:
        for san in moves_san:
            info = engine.analyse(board, limit)
            score = info["score"].relative.score(mate_score=100000)
            relative_cp.append(score if score is not None else 0)
            pv = info.get("pv")
            best_san = None
            if pv:
                try:
                    best_san = board.san(pv[0])
                except ValueError:
                    best_san = pv[0].uci()
            best_moves_san.append(best_san)
            board.push_san(san)

        # final position, no move to classify but needed as the "after" anchor
        info = engine.analyse(board, limit)
        score = info["score"].relative.score(mate_score=100000)
        relative_cp.append(score if score is not None else 0)

    evals = []
    for ply, san in enumerate(moves_san):
        mover = "white" if ply % 2 == 0 else "black"
        value_before = relative_cp[ply]
        value_after = -relative_cp[ply + 1]
        loss = max(0, value_before - value_after)
        clock_seconds = None
        if clocks_centis and ply < len(clocks_centis):
            clock_seconds = clocks_centis[ply] / 100.0
        evals.append(
            {
                "game_id": None,  # filled in by caller
                "ply": ply,
                "mover": mover,
                "move_san": san,
                "cp": value_after if abs(value_after) < 90000 else None,
                "mate": None,
                "best_move": best_moves_san[ply],
                "classification": classify_loss(loss),
                "clock_seconds": clock_seconds,
            }
        )
    return evals

File: client/src/lib/analysis.ts
/** In-browser Stockfish analysis — the client-side replacement for the
 * old server's analysis.py. Runs a full game through the WASM engine and
 * classifies each of the user's moves by centipawn loss. */
import { Chess } from "chess.js";
import { StockfishEngine } from "./engine";
import type { MoveEval } from "./types";

const MATE_SCORE = 100000;

export function classifyLoss(lossCp: number): string {
  if (lossCp >= 300) return "blunder";
  if (lossCp >= 120) return "mistake";
  if (lossCp >= 40) return "inaccuracy";
  if (lossCp <= 5) return "best";
  return "good";
}

// Converts an engine eval (relative to the side to move) into a mate-aware
// numeric score, still relative to the side to move.
function toRelativeCp(ev: { cp: number | null; mate: number | null }): number {
  if (ev.mate !== null) return ev.mate > 0 ? MATE_SCORE - ev.mate : -MATE_SCORE - ev.mate;
  return ev.cp ?? 0;
}

function uciToSan(fen: string, uci: string | null): string | null {
  if (!uci) return null;
  try {
    const chess = new Chess(fen);
    const move = chess.move({
      from: uci.slice(0, 2),
      to: uci.slice(2, 4),
      promotion: uci.length > 4 ? uci.slice(4, 5) : undefined,
    });
    return move?.san ?? uci;
  } catch {
    return uci;
  }
}

export async function analyzeMovesSan(
  movesSan: string[],
  clocksCentis: number[] | null,
  depth: number,
  onProgress?: (done: number, total: number) => void,
): Promise<Omit<MoveEval, "game_id">[]> {
  const engine = new StockfishEngine();
  const chess = new Chess();

  // whiteCp[i] = White-perspective eval of the position *before* move i is
  // played (i.e. after i plies). whiteCp[movesSan.length] is the eval of
  // the final position. Keeping everything in one consistent perspective
  // (rather than "whoever's turn it is") means it can be diffed directly
  // across plies, and matches the convention Lichess's own analysis uses.
  const whiteCp: number[] = [];
  const bestMovesSan: (string | null)[] = [];

  function pushWhiteCp(fen: string, relative: number) {
    const sideToMove = fen.split(" ")[1];
    whiteCp.push(sideToMove === "w" ? relative : -relative);
  }

  try {
    for (let i = 0; i < movesSan.length; i++) {
      const fen = chess.fen();
      const ev = await engine.analyzePosition(fen, depth);
      pushWhiteCp(fen, toRelativeCp(ev));
      bestMovesSan.push(uciToSan(fen, ev.bestMoveUci));
      chess.move(movesSan[i]);
      onProgress?.(i + 1, movesSan.length);
    }
    // A checkmate/stalemate position has no legal moves, so the engine
    // returns no "info" line to score it — special-case it like
    // python-chess's analyse() does, instead of silently treating it as 0.
    if (chess.isCheckmate()) {
      const fen = chess.fen();
      const sideToMove = fen.split(" ")[1];
      // The side to move is checkmated, so it's maximally bad for them.
      whiteCp.push(sideToMove === "w" ? -MATE_SCORE : MATE_SCORE);
    } else if (chess.isGameOver()) {
      whiteCp.push(0);
    } else {
      const fen = chess.fen();
      const finalEv = await engine.analyzePosition(fen, depth);
      pushWhiteCp(fen, toRelativeCp(finalEv));
    }
  } finally {
    engine.terminate();
  }

  const evals: Omit<MoveEval, "game_id">[] = [];
  for (let ply = 0; ply < movesSan.length; ply++) {
    const mover: "white" | "black" = ply % 2 === 0 ? "white" : "black";
    const before = whiteCp[ply];
    const after = whiteCp[ply + 1];
    // Loss is always from the mover's own point of view (how much worse
    // this move made things for them), regardless of which side moved.
    const loss = mover === "white" ? Math.max(0, before - after) : Math.max(0, after - before);
    evals.push({
      ply,
      mover,
      move_san: movesSan[ply],
      // Stored White-perspective, same convention as the Lichess-analysis
      // path — so the eval graph and any diffing across plies is consistent
      // regardless of which engine/source produced the numbers.
      cp: Math.abs(after) < MATE_SCORE - 1000 ? after : null,
      mate: null,
      best_move: bestMovesSan[ply],
      classification: classifyLoss(loss),
      clock_seconds: clocksCentis && ply < clocksCentis.length ? clocksCentis[ply] / 100 : null,
    });
  }
  return evals;
}
